opc-ua-sdk/sdk-server/src/main/java/org/eclipse/milo/opcua/sdk/server/services/DiscoveryServices.java

```diff
@@ -0,0 +1,345 @@
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.Timer;
+import java.util.TimerTask;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
+import java.util.stream.Stream;
+
+import org.eclipse.milo.opcua.sdk.server.api.config.OpcUaServerConfig;
+import org.eclipse.milo.opcua.sdk.server.services.helpers.MdnsHelper;
+import org.eclipse.milo.opcua.stack.core.Identifiers;
+import org.eclipse.milo.opcua.stack.core.StatusCodes;
+import org.eclipse.milo.opcua.stack.core.UaException;
+import org.eclipse.milo.opcua.stack.core.application.services.ServiceRequest;
+import org.eclipse.milo.opcua.stack.core.types.builtin.DateTime;
+import org.eclipse.milo.opcua.stack.core.types.builtin.DiagnosticInfo;
+import org.eclipse.milo.opcua.stack.core.types.builtin.ExtensionObject;
+import org.eclipse.milo.opcua.stack.core.types.builtin.StatusCode;
+import org.eclipse.milo.opcua.stack.core.types.builtin.unsigned.UInteger;
+import org.eclipse.milo.opcua.stack.core.types.structured.FindServersOnNetworkRequest;
+import org.eclipse.milo.opcua.stack.core.types.structured.FindServersOnNetworkResponse;
+import org.eclipse.milo.opcua.stack.core.types.structured.MdnsDiscoveryConfiguration;
+import org.eclipse.milo.opcua.stack.core.types.structured.RegisterServer2Request;
+import org.eclipse.milo.opcua.stack.core.types.structured.RegisterServer2Response;
+import org.eclipse.milo.opcua.stack.core.types.structured.RegisterServerRequest;
+import org.eclipse.milo.opcua.stack.core.types.structured.RegisterServerResponse;
+import org.eclipse.milo.opcua.stack.core.types.structured.RegisteredServer;
+import org.eclipse.milo.opcua.stack.core.types.structured.ResponseHeader;
+import org.eclipse.milo.opcua.stack.core.types.structured.ServerOnNetwork;
+import org.eclipse.milo.opcua.stack.server.tcp.DefaultDiscoveryService;
+import org.eclipse.milo.opcua.stack.server.tcp.UaTcpStackServer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class DiscoveryServices extends DefaultDiscoveryService {
+
+    private LinkedList<RegisteredServer> registeredServers;
+    private HashMap<RegisteredServer, Date> registeredServerLastSeen;
+
+    private final boolean multicastEnabled;
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+
+    private MdnsHelper mdnsHelper = null;
+
+
+    private Consumer<RegisteredServer> registerServerConsumer = null;
+
+    public DiscoveryServices(UaTcpStackServer server, boolean multicastEnabled) {
+        super(server);
+        registeredServers = new LinkedList<>();
+        registeredServerLastSeen = new HashMap<>();
+        this.multicastEnabled = multicastEnabled;
+        new Timer().scheduleAtFixedRate(new TimerTask() {
```

Comment:
Use `Stack.sharedScheduledExecutor()` here instead of a new Timer instance? Not sure how this timer instance could be shutdown and cleaned up when the server and stack get shutdown otherwise.